File: gateway/app/routes/review_queue.py
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel

from app.queries import load
from app.routes.quiz import QuizOption

router = APIRouter()

# No auth system (CLAUDE.md: "one hardcoded student, one instructor") — a
# fixed string is the honest equivalent of a real approved_by identity here.
INSTRUCTOR_ID = "instructor"


class ReviewItem(BaseModel):
    id: str
    lecture_id: str
    lecture_title: str
    concept_id: str | None
    concept_label: str | None
    prompt: str
    options: list[QuizOption]
    correct_option_id: str
    explanation: str | None
    source_timestamp_ms: int
    segment_text: str | None
    frame_id: str | None
    frame_thumb_url: str | None


class ApproveByConceptRequest(BaseModel):
    concept_id: str


@router.get("/review-queue")
async def get_review_queue(request: Request, course_id: str) -> list[ReviewItem]:
    async with request.app.state.pool.acquire() as conn:
        rows = await conn.fetch(load("review_queue.sql"), course_id)
    return [
        ReviewItem(
            id=row["id"],
            lecture_id=row["lecture_id"],
            lecture_title=row["lecture_title"],
            concept_id=row["concept_id"],
            concept_label=row["concept_label"],
            prompt=row["prompt"],
            options=[QuizOption(**o) for o in row["options"]],
            correct_option_id=row["correct_option_id"],
            explanation=row["explanation"],
            source_timestamp_ms=row["source_timestamp_ms"],
            segment_text=row["segment_text"],
            frame_id=row["frame_id"],
            frame_thumb_url=row["frame_thumb_path"],
        )
        for row in rows
    ]


@router.post("/questions/{question_id}/approve")
async def approve_question(request: Request, question_id: str) -> dict:
    async with request.app.state.pool.acquire() as conn:
        result = await conn.execute(load("approve_question.sql"), question_id, INSTRUCTOR_ID)
    if result == "UPDATE 0":
        raise HTTPException(status_code=404, detail=f"No question with id {question_id!r}.")
    return {"id": question_id, "approved": True}


@router.post("/questions/{question_id}/reject")
async def reject_question(request: Request, question_id: str) -> dict:
    async with request.app.state.pool.acquire() as conn:
        result = await conn.execute(load("reject_question.sql"), question_id)
    if result == "DELETE 0":
        raise HTTPException(status_code=404, detail=f"No question with id {question_id!r}.")
    return {"id": question_id, "rejected": True}


@router.post("/review-queue/approve")
async def approve_by_concept(request: Request, body: ApproveByConceptRequest) -> dict:
    async with request.app.state.pool.acquire() as conn:
        rows = await conn.fetch(load("approve_questions_by_concept.sql"), body.concept_id, INSTRUCTOR_ID)
    return {"concept_id": body.concept_id, "approved_ids": [r["id"] for r in rows]}

File: web/src/api/reviewQueue.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiFetch } from './http';
import type { QuizOption } from './quiz';

export interface ReviewItem {
  id: string;
  lecture_id: string;
  lecture_title: string;
  concept_id: string | null;
  concept_label: string | null;
  prompt: string;
  options: QuizOption[];
  correct_option_id: string;
  explanation: string | null;
  source_timestamp_ms: number;
  segment_text: string | null;
  frame_id: string | null;
  frame_thumb_url: string | null;
}

async function fetchReviewQueue(courseId: string): Promise<ReviewItem[]> {
  const res = await apiFetch(`/api/v1/review-queue?course_id=${courseId}`);
  if (!res.ok) throw new Error(`Failed to fetch review queue: ${res.status}`);
  return res.json();
}

export function useReviewQueue(courseId: string) {
  return useQuery({
    queryKey: ['review-queue', courseId],
    queryFn: () => fetchReviewQueue(courseId),
  });
}

function useInvalidateReviewQueue() {
  const queryClient = useQueryClient();
  return () => queryClient.invalidateQueries({ queryKey: ['review-queue'] });
}

export function useApproveQuestion() {
  const invalidate = useInvalidateReviewQueue();
  return useMutation({
    mutationFn: async (questionId: string) => {
      const res = await apiFetch(`/api/v1/questions/${questionId}/approve`, { method: 'POST' });
      if (!res.ok) throw new Error(`Failed to approve: ${res.status}`);
      return res.json();
    },
    onSuccess: invalidate,
  });
}

export function useRejectQuestion() {
  const invalidate = useInvalidateReviewQueue();
  return useMutation({
    mutationFn: async (questionId: string) => {
      const res = await apiFetch(`/api/v1/questions/${questionId}/reject`, { method: 'POST' });
      if (!res.ok) throw new Error(`Failed to reject: ${res.status}`);
      return res.json();
    },
    onSuccess: invalidate,
  });
}

export function useApproveByConcept() {
  const invalidate = useInvalidateReviewQueue();
  return useMutation({
    mutationFn: async (conceptId: string) => {
      const res = await apiFetch('/api/v1/review-queue/approve', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ concept_id: conceptId }),
      });
      if (!res.ok) throw new Error(`Failed to bulk approve: ${res.status}`);
      return res.json();
    },
    onSuccess: invalidate,
  });
}
